### PR TITLE
Support BouncyCastle FIPS JSSE

### LINKED ISSUE #11489: Add optional FIPS mode
**Description**
As part of choosing framework to work with, some organisations need to be FIPS complaint with their products. Quarkus has no FIPS mode. A FIPS mode will be a great ability and will make Quarkus the easiest choice when it comes to choosing framework.

**Links**
[Zulip chat](https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/Fips.20complaint)

### PR DESCRIPTION
**Pull request by sberyozkin**
Fixes #11489

This is the last round for the BouncyCastle provider related work based on the reflective instantiation of the providers until we decide to move in with the dedicated extensions (with a smallprint below :-) ).
I've made the MTLS test working today and what is important there it verifies the Vert.x 3.7.5 fix made by Julien Viet and Paulo, to do with the custom store types/providers (see `VertxHttpRecorder` and the test `application.properties` as well as the test client code which uses Vertx WebClient Mutiny).

The small print is that I've realized I have not added BC related integration tests to the native CI test suite. So I've added some of them, `integration-tests/bouncycastle` and `integration-tests/bouncycastle-jsse` for now, as the new test in this PR, `integration-tests/bouncycastle-fips-jsse` and the FIPS one (strandalone, no JSSE FIPS) I added before, `integration-tests/bouncycastle-fips`, do not work in the native mode yet.

I'd like to propose to try to get this PR before `CR1` as the Vertx related updates are quite important and we can also get the users and SMEs giving it a try; FIPS in the native mode is not critical at this point of time - but I may be able to fix it for one of the 11.0.x final or later releases.

Also CC @cescoffier - hi Clement, FYI, there is a minor mis-alignment for the Vertx Buffer package in the Vertx WebClient code (see the test code, `KeyStoreOptions` and `HttpResponse` `Buffer` take diff packages - fine with me :-), but may be it can be improved somehow)

### CODE REVIEW

File: docs/src/main/asciidoc/security-customization.adoc [hunk 1]
```diff
@@ -287,19 +292,59 @@ If you need to register an `org.bouncycastle.jcajce.provider.BouncyCastleFipsPro
...
+quarkus.http.ssl.certificate.key-store-password=password
+quarkus.http.ssl.certificate.key-store-file-type=BCFKS
+quarkus.http.ssl.certificate.key-store-provider=BCFIPS
+quarkus.http.ssl.certificate.trust-store-file=server-truststore.jks
+quarkus.http.ssl.certificate.trust-store-password=password
+quarkus.http.ssl.certificate.trust-store-file-type=BCFKS
+quarkus.http.ssl.certificate.trust-store-provider=BCFIPS
+----
+
+and the BouncyCastle TLS dependency optimized for using the BouncyCastle FIPS provider:
```
Comment (gastaldi) [inline docs/src/main/asciidoc/security-customization.adoc:326]:
```suggestion
and add the BouncyCastle TLS dependency optimized for using the BouncyCastle FIPS provider:
```

**Review by gastaldi — APPROVED** [review]
LGTM

**Comment by sberyozkin** [comment]
@gastaldi Thanks for that doc fix :-)

**Review by cescoffier — APPROVED** [review]

**Comment by sberyozkin** [comment]
Hi All, this is strange, the native `bouncycastle-jsse` test fails to pick up the `target/quarkus.log`, and indeed in the CI build log I'm seeing `-Dquarkus.log.file.path=target/target/quarkus.log`, I've just rebuilt it locally with `20.3.0`, the native test is good. The test actually checks `target/quarkus.log` and I can see `quarkus.log` in 2 locations:
```
target/target/quarkus.log
target/quarkus.log
```
Also the build log shows a format different to what is configured in the test `application.properties`, which is 
```
quarkus.log.file.format=%C - %s%n
```

for example:

```
org.bouncycastle.jsse.provider.ProvTlsServer - Server selected protocol version: TLSv1.2
```

while in the CI build log it is:

```
2021-01-04 18:23:27,108 FINE  [org.bou.jss.pro.ProvTlsServer] (vert.x-eventloop-thread-0) Server selected protocol version: TLSv1.3
```

the TLS version is different (I'll add one more check) but the configured format is ignored.

**Comment by sberyozkin** [comment]
I've updated the `bouncycastle-jsse` to check both `target/quarkus.log` and `target/target/quarkus.log`, lets see how it goes

**Comment by sberyozkin** [comment]
My last update did help the test find the log file (the test now checks it in two locatiions - looks like the log file can be created too late in the native image as well sometimes), but it could not find the matching records while the MTLS test itself passed. I'm going to do one more tweak and see if it helps to bypass the fact that the log format configuration is apparently not picked up, if it won't work then I'll just skip the log check for this test but in the natve mode only, this test is not even related to this PR :-)

**Comment by sberyozkin** [comment]
For some reasons, the client records are not captured in the log. I've rerun the test locally in the JVM and native mode and it passes with this (trimmed) log output:

1. JVM
```
INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running io.quarkus.it.bouncycastle.BouncyCastleJsseTestCase

2021-01-05 13:33:39,416 INFO  [io.quarkus] (main) Quarkus 999-SNAPSHOT on JVM started in 2.134s. Listening on: http://localhost:8081 and https://localhost:8444
2021-01-05 13:33:39,417 INFO  [io.quarkus] (main) Profile test activated. 
2021-01-05 13:33:39,418 INFO  [io.quarkus] (main) Installed features: [cdi, resteasy, security]
2021-01-05 13:33:40,177 INFO  [org.bou.jss.pro.ProvTrustManagerFactorySpi] (main) Initializing with trust store at path: /home/sberyozkin/work/graalvm-ce-java11-20.3.0/lib/security/cacerts
...
2021-01-05 13:33:40,363 FINE  [org.bou.jss.pro.ProvTlsServer] (vert.x-eventloop-thread-2) Server selected protocol version: TLSv1.2
2021-01-05 13:33:40,451 FINE  [org.bou.jss.pro.ProvTlsClient] (main) Client notified of selected protocol version: TLSv1.2
...
2021-01-05 13:33:40,472 FINE  [org.bou.jss.pro.ProvTlsClient] (main) Client (1.2) selected credentials for signature scheme 'rsa_pss_rsae_sha256(0x804)' (keyType 'RSA'), with private key algorithm 'RSA'

[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 5.226 s - in io.quarkus.it.bouncycastle.BouncyCastleJsseTestCase
2021-01-05 13:33:40,960 FINE  [org.bou.jss.pro.ProvTlsServer] (vert.x-eventloop-thread-2) Server raised warning(1) close_notify(0) alert: Connection closed
2021-01-05 13:33:40,988 INFO  [io.quarkus] (main) Quarkus stopped in 0.123s
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
```
2. Native
```
[INFO] 
[INFO] --- quarkus-maven-plugin:999-SNAPSHOT:build (default) @ quarkus-integration-test-bouncycastle-jsse ---
[INFO] [org.jboss.threads] JBoss Threads version 3.2.0.Final
...
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] Running Quarkus native-image plugin on GraalVM Version 20.3.0 (Java Version 11.0.9+10-jvmci-20.3-b06)
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] /home/sberyozkin/work/graalvm-ce-java11-20.3.0/bin/native-image -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-Dsun.nio.ch.maxUpdateArraySize=100 -J-Dvertx.logger-delegate-factory-class-name=io.quarkus.vertx.core.runtime.VertxLogDelegateFactory -J-Dvertx.disableDnsResolver=true -J-Dio.netty.leakDetection.level=DISABLED -J-Dio.netty.allocator.maxOrder=1 -J-Duser.language=en -J-Dfile.encoding=UTF-8 -H:IncludeResources=.*.jks --initialize-at-build-time= -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy\$BySpaceAndTime -H:+JNI -jar quarkus-integration-test-bouncycastle-jsse-999-SNAPSHOT-runner.jar -H:FallbackThreshold=0 -H:+ReportExceptionStackTraces -H:-AddAllCharsets -H:EnableURLProtocols=http,https --enable-all-security-services -H:NativeLinkerOption=-no-pie --no-server -H:-UseServiceLoaderFeature -H:+StackTrace quarkus-integration-test-bouncycastle-jsse-999-SNAPSHOT-runner
[quarkus-integration-test-bouncycastle-jsse-999-SNAPSHOT-runner:28678]    classlist:   2,930.65 ms,  0.96 GB
[quarkus-integration-test-bouncycastle-jsse-999-SNAPSHOT-runner:28678]        (cap):     591.90 ms,  0.96 GB
[quarkus-integration-test-bouncycastle-jsse-999-SNAPSHOT-runner:28678]        setup:   1,851.10 ms,  0.96 GB

[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running io.quarkus.it.bouncycastle.BouncyCastleJsseITCase
Executing [/home/sberyozkin/work/rh/quarkusio/quarkus/integration-tests/bouncycastle-jsse/target/quarkus-integration-test-bouncycastle-jsse-999-SNAPSHOT-runner, -Dquarkus.http.port=8081, -Dquarkus.http.ssl-port=8444, -Dtest.url=http://localhost:8081, -Dquarkus.log.file.path=target/target/quarkus.log]
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2021-01-05 13:35:40,895 INFO  [io.quarkus] (main) quarkus-integration-test-bouncycastle-jsse 999-SNAPSHOT native (powered by Quarkus 999-SNAPSHOT) started in 0.012s. Listening on: http://0.0.0.0:8081 and https://0.0.0.0:8444
2021-01-05 13:35:40,895 INFO  [io.quarkus] (main) Profile prod activated. 
2021-01-05 13:35:40,895 INFO  [io.quarkus] (main) Installed features: [cdi, resteasy, security]

2021-01-05 13:35:42,442 FINE  [org.bou.jss.pro.ProvTlsServer] (vert.x-eventloop-thread-0) Server selected protocol version: TLSv1.3

[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.983 s - in io.quarkus.it.bouncycastle.BouncyCastleJsseITCase
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
```

So I'll just disable checking the log for the native mode as it is not quite reliable

**Comment by sberyozkin** [comment]
OK, so indeed in CI the log for this test does not contain the client TLS records in the native image - but the actual MTLS test works so I'll be happy with it - at least I'll update the test to check the server records in the native image and it will be fine

**Comment by sberyozkin** [comment]
File log checks are not really reliable in CI - it appears some async logging is in place. Also I saw a message that the BC log level (as in the application.properties, set to TRACE) is below the minimum required INFO level - not sure what does it mean. Anyway, I'm only keeping now the log check for the BC_JSSE test - both client and server records are checked in the JVM mode and the server records in the native mode... I'd just like to get the VertHttpRecorder changes in, the FIPS JSSE MTLS test involvinh the actual connection with the BC-FIPS formatted stores works fine

**Comment by sberyozkin** [comment]
I'm going to go ahead and will follow up with opening an issue about the FIPS tests in the native mode

**Pull request merged**